GH-98906 ```re``` module: ```search() vs. match()``` section should mention ```fullmatch()```

ramvikrams opened the pull request:
Added the re.fullmatch() with it's examples

GH-98906 ```re``` module: ```search() vs. match()``` section should mention ```fullmatch()```

ramvikrams committed:
for #98906 

adding re.fullmatch with it's examples

slateny commented on Doc/library/re.rst:
```diff
@@ -1565,18 +1565,22 @@ search() vs. match()
 Python offers two different primitive operations based on regular expressions:
```
Maybe continue the style of this paragraph, so instead something like 'Python offers a few primitive ...' then '... re.fullmatch checks that the whole string is a match ...' (can change this wording up a bit too). I also think the `None` case can omitted since it wasn't mentioned in the previous two cases either.

ramvikrams replied:
Done sir

slateny commented on Doc/library/re.rst:
```diff
@@ -1586,6 +1590,7 @@ beginning of the string, whereas using :func:`search` with a regular expression
 beginning with ``'^'`` will match at the beginning of each line. ::
 
    >>> re.match('X', 'A\nB\nX', re.MULTILINE)  # No match
```
For consistency, other examples seem to use double quotes so this and `re.search` can be swapped over, or `re.fullmatch` can be changed to single quote (but optional).

ramvikrams replied:
Done the changes

ramvikrams committed:
changes

ramvikrams committed:
update

slateny commented on Doc/library/re.rst:
```diff
@@ -1562,21 +1562,24 @@ search() vs. match()
+Python offers different primitive operations based on regular expressions:
+:func:`re.match` checks for a match only at the beginning of the string,
+:func:`re.fullmatch` checks for entire string to be a match. :func:`re.search`
+checks for a match anywhere in the string (this is what Perl does by default).
```
For better flow

```suggestion
:func:`re.fullmatch` checks for entire string to be a match, and
:func:`re.search` checks for a match anywhere in the string
(this is what Perl does by default).
```

or perhaps even

```suggestion
:func:`re.fullmatch` checks for entire string to be a match, and
:func:`re.search` checks for a match anywhere in the string.
```

but I'll let Eric decide whether the Perl mention is necessary.

ramvikrams replied:
ok

hauntsaninja replied:
My recommendation would be to use bullet points (and to reorder so that plain `re.match` is last, honestly it's a little bit of a footgun)
```
* :func:`re.search` checks for a match anywhere in the string
  (this is what Perl does by default)
* :func:`re.fullmatch` checks for entire string to be a match
* :func:`re.match` checks for a match only at the beginning of the string
```

hauntsaninja commented on Doc/library/re.rst:
```diff
@@ -1562,21 +1562,24 @@ search() vs. match()
    >>> re.search("c", "abcdef")   # Match
    <re.Match object; span=(2, 3), match='c'>
+   >>> re.fullmatch("python", "python")  # Match
+   <re.Match object; span=(0, 6), match='python'>
```
None of these examples differentiate between `re.fullmatch` and `re.match`

ramvikrams replied:
Thanks I'll do it

ramvikrams replied:
what example can we take then to differentiate between them

ericvsmith replied:
I think you'd need to show an example where `search()` or `match()` match, but `fullmatch()` does not.

ramvikrams replied:
ok, i'll do that

ramvikrams replied:
done with the changes, sir

ramvikrams committed:
update

ramvikrams committed:
Merge branch 'python:main' into t2

ramvikrams committed:
update the examples 

update the examples  where `re.match` and `re.search` shows a match but not `re.fullmatch`

ericvsmith commented on Doc/library/re.rst:
```diff
@@ -1565,21 +1565,27 @@ search() vs. match()
 
-   >>> re.match("c", "abcdef")    # No match
-   >>> re.search("c", "abcdef")   # Match
-   <re.Match object; span=(2, 3), match='c'>
```
I don't think there was anything wrong with the first 3 lines. I suggest leaving them as-is.

ramvikrams replied:
yes sir

ericvsmith commented on Doc/library/re.rst:
```diff
@@ -1565,21 +1565,27 @@ search() vs. match()
+   <re.Match object; span=(0, 1), match='c'>
+   >>> re.search("c", "cdef")   # Match
+   <re.Match object; span=(0, 1), match='c'>
+   >>> re.fullmatch("c", "cdef")  # No Match
```
Sorry I wasn't clear. I think you need two examples for fullmatch:
```python
re.fullmatch("p.*n", "python")  # Match
re.fullmatch("r.*n", "python") # No match
```

ramvikrams replied:
yes sir

ramvikrams replied:
done sir

ericvsmith commented on Doc/library/re.rst:
```diff
@@ -1565,21 +1565,27 @@ search() vs. match()
+   >>> re.match("c", "cdef")    # match
+   <re.Match object; span=(0, 1), match='c'>
+   >>> re.search("c", "cdef")   # Match
+   <re.Match object; span=(0, 1), match='c'>
```
And I would not add these 2 examples. The existing examples are fine. The goal is to show that `search()` can match something that `match()` does not, with the same parameters to both.

ramvikrams replied:
yes sir

ericvsmith commented on Doc/library/re.rst:
```diff
@@ -1565,21 +1565,27 @@ search() vs. match()
 restrict the match at the beginning of the string::
 
    >>> re.match("c", "abcdef")    # No match
+   >>> re.fullmatch("c", "abcdef")  # No Match
```
I would not add this. This section is trying to show the differences between `search` and `match`. Adding a `fullmatch` example just confuses things.

ramvikrams replied:
yes sir

ericvsmith replied:
Please delete the `fullmatch` line here.

ramvikrams committed:
updated the docs

ericvsmith commented on Doc/library/re.rst:
```diff
@@ -1588,8 +1595,9 @@ Note however that in :const:`MULTILINE` mode :func:`match` only matches at the
-   >>> re.match('X', 'A\nB\nX', re.MULTILINE)  # No match
-   >>> re.search('^X', 'A\nB\nX', re.MULTILINE)  # Match
+   >>> re.match("X", "A\nB\nX", re.MULTILINE)  # No match
+   >>> re.fullmatch("X", "A\nB\nX", re.MULTILINE)  # No Match
```
And delete this`fullmatch` line also. This section doesn't need to mention fullmatch.

ericvsmith commented on Doc/library/re.rst:
```diff
@@ -1565,21 +1565,28 @@ search() vs. match()
++ :func:`re.search` checks for a match anywhere in the string,
+  (this is what Perl does by default)
++ :func:`re.fullmatch` checks for entire string to be a match
++ :func:`re.match` checks for a match only at the beginning of the string
```
Please leave these in the order of `match`, then `search`, then `fullmatch`. That way the match the examples immediately below.

ramvikrams committed:
Update re.rst

ericvsmith commented on Doc/library/re.rst:
```diff
@@ -1565,16 +1565,22 @@ search() vs. match()
    >>> re.match("c", "abcdef")    # No match
    >>> re.search("c", "abcdef")   # Match
    <re.Match object; span=(2, 3), match='c'>
+   >>> re.fullmatch("p.*n", "python")
```
Sorry I missed this: Can you add the "# Match" comment to this line?

ericvsmith reviewed (changes requested):
Sorry for one additional small change. I'll commit it after this change.

ramvikrams committed:
Update re.rst
